src/plugins/intel_gpu/src/graph/impls/common/loop.cpp

```diff
@@ -37,63 +107,173 @@ struct loop_impl : typed_primitive_impl<loop> {
+        if (!primitive->first_execution_condition_id.empty()) {
+            memory::ptr first_execution_condition_mem = outer_network.get_primitive(primitive->first_execution_condition_id)->output_memory_ptr();
+            execution_condition = read_scalar_value(first_execution_condition_mem, stream);
+        }
+
+        // When execution_condition is false or trip_count is zero, return execute_impl without any body_network execution.
+        if (!execution_condition || trip_count == 0) {
+            // Update actual num iteration
+            if (!primitive->num_iteration_id.empty()) {
+                // update num_iterations (actual number of iterations)
+                memory::ptr num_actual_iterations_mem = outer_network.get_primitive(primitive->num_iteration_id)->output_memory_ptr();
+                write_scalar_value(num_actual_iterations_mem, stream, current_iteration_idx);
             }
-            instance.preproc_memories_done = true;
+
+            instance.update_output_layout();
+            ev->set();
+            return ev;
         }
 
-        // read trip_count from outer network
-        bool update_num_iterations = false;
-        memory::ptr trip_count_mem = outer_network.get_primitive(primitive->trip_count_id)->output_memory_ptr();
-        int64_t trip_count = loop_node::read_scalar_value(std::move(trip_count_mem), stream);
-        if (trip_count < 0) {
-            trip_count = _max_iteration;
-            update_num_iterations = true;
+        //////////////////////////////////////////
+        // memory pointers for body network
+        //////////////////////////////////////////
+        // shortcut of execution_condition memory in body network
+        memory::ptr body_execution_condition_mem = nullptr;
+        if (!primitive->body_execution_condition_id.empty()) {
+            body_execution_condition_mem = body_network->get_primitive(primitive->body_execution_condition_id)->output_memory_ptr();
         }
 
-        // read initial execution condition from outer network
-        memory::ptr initial_execution_mem = outer_network.get_primitive(primitive->initial_execution_id)->output_memory_ptr();
-        int64_t execution_condition = loop_node::read_scalar_value(initial_execution_mem, stream);
+        // shortcut of current_iteration memory in body network
+        memory::ptr body_current_iteration_mem = nullptr;
+        if (!primitive->body_current_iteration_id.empty()) {
+            body_current_iteration_mem = body_network->get_primitive(primitive->body_current_iteration_id)->output_memory_ptr();
+        }
 
-        // shortcut of execution_condition memory in body network
-        memory::ptr execution_condition_mem = nullptr;
-        if (!primitive->condition_id.empty()) {
-            execution_condition_mem = body_network->get_primitive(primitive->condition_id)->output_memory_ptr();
+        if (instance.is_dynamic()) {
+            instance.update_shape();
```

Comment:
Isnt this done already before this point??

Reply:
It does not update shape before this point because we skip the update shape process in primitive_inst if current node has inner network.